connectors/src/connectors/microsoft/temporal/activities.ts

```diff
@@ -819,3 +698,164 @@ async function handleTextExtraction(
       }
     : null;
 }
+
+async function upsertDocument(
+  dataSourceConfig: DataSourceConfig,
+  file: microsoftgraph.DriveItem,
+  documentContent: CoreAPIDataSourceDocumentSection | null,
+  documentId: string,
+  maxDocumentLen: number,
+  localLogger: Logger,
+  connectorId: ModelId,
+  startSyncTs: number,
+  isBatchSync: boolean,
+  parentInternalId: string
+): Promise<number | undefined> {
+  const updatedAt = file.lastModifiedDateTime
+    ? new Date(file.lastModifiedDateTime)
+    : undefined;
+
+  const createdAt = file.createdDateTime
+    ? new Date(file.createdDateTime)
+    : undefined;
+
+  const content = await renderDocumentTitleAndContent({
+    dataSourceConfig,
+    title: file.name ?? null,
+    updatedAt,
+    createdAt: createdAt,
+    lastEditor: file.lastModifiedBy?.user?.displayName ?? undefined,
+    content: documentContent,
+  });
+
+  if (documentContent === undefined) {
+    localLogger.error({}, "documentContent is undefined");
+    throw new Error("documentContent is undefined");
+  }
+
+  const tags = [`title:${file.name}`];
+  if (updatedAt) {
+    tags.push(`updatedAt:${updatedAt}`);
+  }
+  if (file.createdDateTime) {
+    tags.push(`createdAt:${file.createdDateTime}`);
+  }
+  if (file.lastModifiedBy?.user?.displayName) {
+    tags.push(`lastEditor:${file.lastModifiedBy.user.displayName}`);
+  }
+
+  tags.push(`mimeType:${file.file?.mimeType}`);
+
+  const documentLength = documentContent ? sectionLength(documentContent) : 0;
+  const upsertTimestampMs = updatedAt ? updatedAt.getTime() : undefined;
+  const isInSizeRange = documentLength > 0 && documentLength < maxDocumentLen;
+  if (isInSizeRange) {
+    const parents = await getParents({
+      connectorId,
+      internalId: documentId,
+      parentInternalId,
+      startSyncTs,
+    });
+    parents.reverse();
+
+    await upsertToDatasource({
+      dataSourceConfig,
+      documentId,
+      documentContent: content,
+      documentUrl: file.webUrl ?? undefined,
+      timestampMs: upsertTimestampMs,
+      tags,
+      parents: parents,
+      upsertContext: {
+        sync_type: isBatchSync ? "batch" : "incremental",
+      },
+      async: true,
+    });
+  } else {
+    localLogger.info(
+      { documentLength },
+      "Document is empty or too big to be upserted. Skipping"
+    );
+    return undefined;
+  }
+}
+
+/**
+ * Early return if lastSeenTs is greater than workflow start.
+ * This allows avoiding resyncing already-synced documents in case of activity failure
+ */
+export function shouldSkipSync(
```

Comment:
IMO we should not extract in a function. Usually a big fan of separating into function, but here it's too small and makes it counter productive for readability

"shouldSkipSync" is a new word to learn for those who read the code, why and when should we skip sync? there could be many very different reasons and so here when I read I will want to go to the funciton and "oh, it's just that"

While `fileResource?.lastSeenTs && fileResource.lastSeenTs > new Date(startSyncTs)` is quite concise and clear already

If the function was used ~4 or more times, with a clearer name, maybe this would have been worth it